nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,356 @@
+  slowdown_marker.scale.z = 0.02;
+  slowdown_marker.color.a = 0.2;
+  slowdown_marker.color.r = 0.0;
+  slowdown_marker.color.g = 1.0;
+  slowdown_marker.color.b = 0.0;
+  return slowdown_marker;
+}
+
+bool GracefulMotionController::simulateTrajectory(
+  const geometry_msgs::msg::PoseStamped & robot_pose,
+  const geometry_msgs::msg::PoseStamped & motion_target, nav_msgs::msg::Path & trajectory)
+{
+  // Check for cllision before moving
+  if (inCollision(
+      robot_pose.pose.position.x, robot_pose.pose.position.y,
+      tf2::getYaw(robot_pose.pose.orientation)))
+  {
+    return false;
+  }
+
+  // First pose
+  geometry_msgs::msg::PoseStamped next_pose;
+  next_pose.header.frame_id = costmap_ros_->getBaseFrameID();
+  next_pose.pose.orientation.w = 1.0;
+  trajectory.poses.push_back(next_pose);
+
+  double distance = std::numeric_limits<double>::max();
+  double resolution_ = costmap_ros_->getCostmap()->getResolution();
+  double dt = (params_->v_linear_max > 0.0) ? resolution_ / params_->v_linear_max : 0.0;
+
+  // Set max iter to avoid infinite loop
+  unsigned int max_iter = 2 * sqrt(
+    motion_target.pose.position.x * motion_target.pose.position.x +
+    motion_target.pose.position.y * motion_target.pose.position.y) / resolution_;
+
+  // Generate path
+  do{
+    // Apply velocities to calculate next pose
+    next_pose.pose = control_law_->calculateNextPose(dt, motion_target.pose, next_pose.pose);
+
+    // Add the pose to the trajectory for visualization
+    trajectory.poses.push_back(next_pose);
+
+    // Check for collision
+    geometry_msgs::msg::PoseStamped global_pose;
+    tf2::doTransform(next_pose, global_pose, costmap_transform_);
+    if (inCollision(
+        global_pose.pose.position.x, global_pose.pose.position.y,
+        tf2::getYaw(global_pose.pose.orientation)))
+    {
+      return false;
+    }
+
+    // Check if we reach the goal
+    double error_x = motion_target.pose.position.x - next_pose.pose.position.x;
+    double error_y = motion_target.pose.position.y - next_pose.pose.position.y;
+    distance = std::hypot(error_x, error_y);
+  }while(distance > resolution_ && trajectory.poses.size() < max_iter);
+
+  return true;
+}
+
+geometry_msgs::msg::Twist GracefulMotionController::rotateToTarget(const double & angle_to_target)
+{
+  double sign = angle_to_target < 0.0 ? -1.0 : 1.0;
+  geometry_msgs::msg::Twist vel;
+  vel.linear.x = 0.0;
+  vel.angular.z = sign * params_->v_angular_max;
+  return vel;
+}
+
+bool GracefulMotionController::inCollision(const double & x, const double & y, const double & theta)
+{
+  unsigned int mx, my;
+  if (!costmap_ros_->getCostmap()->worldToMap(x, y, mx, my)) {
+    RCLCPP_WARN(
+      logger_, "The path is not in the costmap. Cannot check for collisions. "
+      "Proceed at your own risk, slow the robot, or increase your costmap size.");
+    return false;
+  }
+
+  double footprint_cost = collision_checker_->footprintCostAtPose(
```

Comment:
Its worth having a check if this footprint is a circle, in which a point check is much more efficient than full footprint SE2 check